internal/pkg/service/stream/storage/level/local/diskwriter/router/connection/connection.go

```diff
@@ -0,0 +1,241 @@
+// Package connection manages connections from the current source node to all disk writer nodes.
+//
+// Principle of functionality:
+//   - Each disk writer node reports its own mounted volumes to the database.
+//   - The connection Manager watches for writer volumes changes, so it knows about all on-line disk writer nodes.
+//   - With every change in writers volumes, it is checked that a connection to each writer nodes exists.
+//   - Each connection is independent and covered by infinite retries, see transport.Client.OpenConnection.
+//   - Connections are used by the storage router to send data to individual slices (disk writer nodes).
+package connection
+
+import (
+	"context"
+	"slices"
+	"strings"
+	"sync"
+
+	etcd "go.etcd.io/etcd/client/v3"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/prefixtree"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/servicectx"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/diskwriter/network"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/diskwriter/network/transport"
+	volume "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/volume/model"
+	storageRepo "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/model/repository"
+)
+
+// Manager manages connections from the current source node to all disk writer nodes.
+type Manager struct {
+	logger log.Logger
+
+	// client opens connections to disk writer nodes.
+	client *transport.Client
+
+	// volumes field contains in-memory snapshot of all active disk writer volumes.
+	// It is used to get info about active disk writers, to open/close connections.
+	volumes *etcdop.Mirror[volume.Metadata, *volumeData]
+
+	closed <-chan struct{}
+	wg     sync.WaitGroup
+
+	connectionsLock sync.Mutex
+	connections     map[string]*connection
+}
+
+type volumeData struct {
+	ID   volume.ID
+	Node *nodeData
+}
+
+type nodeData struct {
+	ID      string
+	Address volume.RemoteAddr
+}
+
+type connection struct {
+	Node       *nodeData
+	ClientConn *transport.ClientConnection
+}
+
+type dependencies interface {
+	Logger() log.Logger
+	Process() *servicectx.Process
+	StorageRepository() *storageRepo.Repository
+}
+
+func NewManager(ctx context.Context, d dependencies, cfg network.Config, nodeID string) (*Manager, error) {
+	m := &Manager{
+		logger:      d.Logger().WithComponent("storage.router.connections"),
+		connections: make(map[string]*connection),
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
I don't like much the shadowing, could we rename it ? e.g `managerCtx`

Reply:
@jachym-tousek-keboola 
In my opinion, shadowing is better in this use-case.
Because in the code bellow, you have only one `ctx` variable, and therefore you cannot make a mistake, by using some different context.

Reply:
other approach is to call the context in the parameter `baseContext` for example.